Domains contracts management

vi4m opened the pull request:
## TOC
- Added new Domain Management module for domains purchases tracking.
- Some love to the make flake target - clean stage before flake was _fucking_ slow, and in fact not needed, so I removed it. You can always do this in a explicit way(`make clean flake`). 

From now on, beautiful ascii art flake will show, when `make flake` passes succesfully.
## Domain management Q&A

Q: What services/env will be used for?
A: All domains will have Service/Env set to -> Domains Mangement Service / prod / test. Scrooge will take care of prices management just like "Virtualization Systems"

Q: Why Technical and Business owner are separate from Service's owners. 
A: Because they are just responsible for a domain.

Q: Why separation of Domain and Contracts?
A: 1. Domain will be reused by django-powerdns-*. 2. We want to maintain history in a clean way, and managing not only renewal actions, but other types as well, such as transfer costs (another contract you can add in the future)

mkurek commented on src/ralph/admin/sitetrees.py:
```diff
@@ -123,6 +123,17 @@ def section(section_name, app, model):
+            children=[
+                section(_('Domain contracts'), 'domains', 'DomainContract'),
+                section(_('Domain registrants'), 'domains', 'DomainRegistrant'),
+                section(_('Domain owners'), 'accounts', 'RalphUser'),
```
is this a mistake? what is the purpose of this menu entry, since we have 'Users' already?

vi4m replied:
There are multiple problems with the "Users" sitting somewhere in menu settings. 
1. "Don't let me think". We don't want the users to think about what's underlaying model for Domain owner field (User) and where to find it in settings. The fact that we use the same object is just a technical coincidence IMO. 
2. The fact, that you have to create ralph account(with password) just to assign "Domain owner" is a problem too, but this PR doesn't touch it

mkurek replied:
if you mean Domain Driven Design, then not every user is domain owner (i know it could be, but only specific subset of users actually would be owning some domains), so IMO we should filter this view to show only real domain owners. What do you think?

vi4m replied:
It could be great to have either subclass of User for custom cases, or other "filter" like mechanism. I'm not sure what's better.

mkurek replied:
filter (consider using django proxy model with overwritten db manager (objects): https://docs.djangoproject.com/en/1.8/topics/db/models/#proxy-models)

vi4m replied:
I just left Users table without filters for now, to not complicate things.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
```
since there is only one file with only few models you could use `models.py` directly instead of creating models directory. We were using models directory iff there were lot of models in single app.

vi4m replied:
Good point, if this PR ends up so simple, i will just live one file for models.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,80 @@
+        choices=DomainStatus(),
+        default=DomainStatus.active.id,
+    )
+    service = models.ManyToManyField(Service)
```
consider, if it should point to `ServiceEnvironment` model (`ralph.assets.models.ServiceEnvironment`) - then it'll store information about environment either (at the same time).

vi4m replied:
No, it's revelant only to service as business wanted.

mkurek replied:
i don't know if there are any plans to charge for domains in scrooge, but if there will be such in future then it'll be impossible in current model, since scrooge requires both service and environment.

vi4m replied:
Fixed.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,80 @@
+        default=DomainStatus.active.id,
+    )
+    service = models.ManyToManyField(Service)
+    expiration_date = models.DateField(null=True, blank=True)
```
what you think about storing history of expiration dates (and maybe additional info too, for example some contract id, price etc)?

vi4m replied:
It depends on searching/reporting, it has to be clarified if users prefers explicit dates management, or django history will be sufficient for auditable purposes. Has to be decided after end users tests and expectations.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,80 @@
+    )
+    service = models.ManyToManyField(Service)
+    expiration_date = models.DateField(null=True, blank=True)
+    business_segment = models.ForeignKey(BusinessSegment)
```
business segment is defined indirectly by service (service -> profit center -> business segment)

vi4m replied:
It could be many services assigned to one domain contract, but only one bussines segment is the owner, that's why it's here "again".

mkurek replied:
ok, so shouldn't there be validation if selected services are descendants of selected business segment?

vi4m replied:
I'm not sure if the reality will be so simple.TBD further. (we have 2k records to import with history baggage included and ownerships are not so simple in reality ;-))

vi4m replied:
See the notes in the description field.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,80 @@
+    transfered_away = _('Transfered away')
+
+
+class DomainContract(
```
do we want to store only public domains? If not, then what about distinction between public and private domains? Or this would be stored in django_powerdns_dnssec and related?

vi4m replied:
django_powerdns_dnssec

mkurek commented on src/ralph/domains/filters.py:
```diff
@@ -0,0 +1,16 @@
+# # -*- coding: utf-8 -*-
```
when https://github.com/allegro/ralph/pull/1807 will be merged, these filters won't be necessary

mkurek commented on Makefile:
```diff
@@ -35,7 +35,8 @@ isort:
 
-flake: clean isort
+flake: isort
+	@cat scripts/flake.txt
```
what about this file? why didn't you commited it?

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,98 @@
+        help_text=_("Technical owner of a service")
+    )
+    domain_holder = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
```
this should point to `AssetHolder` model

vi4m committed:
Show beautifull ascii art flake after successfull "make flake"

vi4m committed:
Added domain management system.

This module manages ownership, purpose and price aspects of a domains.
For more information read the docs/domains/overview.md documentation.

mkurek commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,75 @@
+        'domain__name', 'domain__service_env__service__name'
+    ]
+    list_display = [
+        'domain', 'expiration_date', 'price'
```
what about service? if we allow to filter by service name shouldn't it be one of the list columns too?

vi4m replied:
service name search is still present for open source in order to use this field, but table view doesn't show that because in our case it will show "Domain Management Service" for all records.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,94 @@
+        choices=DomainStatus(),
+        default=DomainStatus.active.id,
+    )
+    business_segment = models.ForeignKey(
```
i don't remember if i've asked it before but what do you think about many-to-many here? will there be one (and only one) contact person for domain? In case of services we allow for many owners.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,94 @@
+    active = _('Active')
+    pending_lapse = _('Pending lapse')
+    pending_transfer = _('Pending transfer away')
+    lapsed = _('Lapsed (inactive)')
```
maybe `expired` will better suit here?

vi4m replied:
NO. "lapsed" is used in domain management dictionary. Expired is the process "in the middle"

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,94 @@
+        help_text=_("Business segment for a domain")
+    )
+    business_owner = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
```
i don't remember if i've asked it before but what do you think about many-to-many here? will there be one (and only one) contact person for domain? In case of services we allow for many owners.

vi4m replied:
We don't need this, since it's the person just for a single domain management.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,94 @@
+    )
+
+    def __str__(self):
+        return "{name}".format(
```
just `return self.name` will be enough :)

mkurek commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,75 @@
+        }),
+        (_('Ownership info'), {
+            'fields': (
+                'service_env', 'domain_status',
```
you have duplicated `domain_status` field

mkurek commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,75 @@
+
+@register(Domain)
+class DomainAdmin(RalphAdmin):
+    pass
```
remove this `pass`

mkurek commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,75 @@
+    list_filter = [
+        'name', 'service_env',
+    ]
+    list_display = [
```
what about service_env?

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,94 @@
+        )
+
+
+class DomainContract(
```
what about some asset-like fields like invoice date, invoice number, order number, task url etc?

vi4m replied:
It's automatically handled by auto-renewal process of CNC. If we want it in the future, we will just extend the model.

mkurek commented on src/ralph/domains/models/domains.py:
```diff
@@ -0,0 +1,94 @@
+    transfered_away = _('Transfered away')
+
+
+class Domain(BaseObject):
```
mix `AdminAbsoluteUrlMixin` here (`ralph.lib.mixins.models.AdminAbsoluteUrlMixin`)

mkurek commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,75 @@
+
+
+@register(Domain)
+class DomainAdmin(RalphAdmin):
```
don't we want to use attachments here? if yes, mix `AttachmentsMixin` (`ralph.attachments.admin.AttachmentsMixin`)

vi4m replied:
good point :)

mkurek commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,75 @@
+
+
+@register(DomainContract)
+class DomainContract(RalphAdmin):
```
consider attachments here too

mkurek commented:
:ok:

ar4s commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,76 @@
+
+@register(Domain)
+class DomainAdmin(AttachmentsMixin, RalphAdmin):
+    change_views = []
```
pls remove - https://github.com/allegro/ralph/pull/1944

vi4m replied:
removed

ar4s commented on src/ralph/domains/admin.py:
```diff
@@ -0,0 +1,76 @@
+
+@register(DomainContract)
+class DomainContractAdmin(AttachmentsMixin, RalphAdmin):
+    change_views = []
```
ditto

vi4m replied:
removed

vi4m committed:
Merge branch 'ng' of https://github.com/allegro/ralph into squashed-domains

ar4s commented:
:ok:
